File: 09-Graph/09-Minimum-Spanning-Tree/prim.c
#include <stdio.h>

#define MAX 100
#define INF 999999

int graph[MAX][MAX];

int main() {
    int n, e;

    printf("Enter number of vertices: ");
    scanf("%d", &n);

    for(int i = 0; i < n; i++) {
        for(int j = 0; j < n; j++) {
            graph[i][j] = INF;
        }
    }

    printf("Enter number of edges: ");
    scanf("%d", &e);

    printf("Enter weighted edges (u v w):\n");

    for(int i = 0; i < e; i++) {
        int u, v, w;

        scanf("%d %d %d", &u, &v, &w);

        graph[u][v] = w;
        graph[v][u] = w;
    }

    int start;

    printf("Enter starting vertex: ");
    scanf("%d", &start);

    int key[MAX];
    int parent[MAX];
    int visited[MAX] = {0};

    for(int i = 0; i < n; i++) {
        key[i] = INF;
        parent[i] = -1;
    }

    key[start] = 0;

    int total_cost = 0;
    int edges_used = 0;

    printf("\nEdges in Minimum Spanning Tree:\n");

    for(int count = 0; count < n; count++) {

        int min = INF;
        int node = -1;

        for(int i = 0; i < n; i++) {
            if(!visited[i] && key[i] < min) {
                min = key[i];
                node = i;
            }
        }

        if(node == -1)
            break;

        visited[node] = 1;

        if(parent[node] != -1) {
            printf("%d -- %d  weight: %d\n",
                   parent[node],
                   node,
                   key[node]);

            total_cost += key[node];
            edges_used++;
        }

        for(int i = 0; i < n; i++) {
            if(graph[node][i] != INF &&
               !visited[i] &&
               graph[node][i] < key[i]) {

                key[i] = graph[node][i];
                parent[i] = node;
            }
        }
    }

    if(edges_used != n - 1) {
        printf("\nMST is not possible because the graph is disconnected.\n");
    } else {
        printf("\nMinimum Cost: %d\n", total_cost);
    }

    return 0;
}

File: 09-Graph/09-Minimum-Spanning-Tree/prim.py
import heapq

n = int(input("Enter number of vertices: "))
e = int(input("Enter number of edges: "))

graph = [[] for _ in range(n)]

print("Enter weighted edges (u v w):")

for _ in range(e):
    u, v, w = map(int, input().split())

    graph[u].append((v, w))
    graph[v].append((u, w))

start = int(input("Enter starting vertex: "))

visited = [False] * n
pq = [(0, start, -1)]

total_cost = 0
edges_used = 0

print("\nEdges in Minimum Spanning Tree:")

while pq:

    weight, node, parent = heapq.heappop(pq)

    if visited[node]:
        continue

    visited[node] = True

    if parent != -1:
        print(parent, "--", node, "weight:", weight)
        total_cost += weight
        edges_used += 1

    for neighbor, edge_weight in graph[node]:
        if not visited[neighbor]:
            heapq.heappush(
                pq,
                (edge_weight, neighbor, node)
            )

if edges_used != n - 1:
    print("\nMST is not possible because the graph is disconnected.")
else:
    print("\nMinimum Cost:", total_cost)